crates/data_components/src/duckdb.rs

```diff
@@ -260,24 +281,44 @@ impl DuckDB {
     }
 }
 
-// TODO: DuckDB DataConnector
-//
-// #[async_trait]
-// impl Read for DuckDBTableFactory {
-//     async fn table_provider(
-//         &self,
-//         _table_reference: OwnedTableReference,
-//     ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn std::error::Error + Send + Sync>> {
-//         todo!()
-//     }
-// }
-
-// #[async_trait]
-// impl ReadWrite for DuckDBTableFactory {
-//     async fn table_provider(
-//         &self,
-//         _table_reference: OwnedTableReference,
-//     ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn std::error::Error + Send + Sync>> {
-//         todo!()
-//     }
-// }
+pub struct DuckDBTableFactory {
+    pool: Arc<DuckDbConnectionPool>,
+}
+
+impl DuckDBTableFactory {
+    #[must_use]
+    pub fn new(pool: Arc<DuckDbConnectionPool>) -> Self {
+        Self { pool }
+    }
+}
+
+#[async_trait]
+impl Read for DuckDBTableFactory {
+    async fn table_provider(
+        &self,
+        table_reference: OwnedTableReference,
+    ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn std::error::Error + Send + Sync>> {
+        let pool = Arc::clone(&self.pool);
+        let dyn_pool: Arc<DynDuckDbConnectionPool> = pool;
+        let table_provider = SqlTable::new(&dyn_pool, table_reference)
+            .await
+            .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)?;
+
+        Ok(Arc::new(table_provider))
+    }
+}
+
+#[async_trait]
+impl ReadWrite for DuckDBTableFactory {
+    async fn table_provider(
+        &self,
+        table_reference: OwnedTableReference,
+    ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn std::error::Error + Send + Sync>> {
+        let read_provider = Read::table_provider(self, table_reference.clone()).await?;
+
+        let table_name = table_reference.to_string();
+        let duckdb = DuckDB::new(table_name, Arc::clone(&self.pool));
```

Comment:
The `read_provider` can give you the schema at this point to pass into the `duckdb` instance.

Reply:
Schema retrieved from read_provider in the updated code